.github: add basic release CI

AnnaShaleva opened the pull request:
Close #78, close #103.

In this PR:
- Add binary builder job that always uploads built binaries as artifacts and (on release) attaches them to the Github release.
- Add Docker image builder job that always build Docker image and (on release or request) uploads it to the Dockerhub.

Draft PR, need to:

- [x] Decide about first version number.
- [x] ~~Set first Git tag.~~ Will be done on first release.
- [x] Test.
- [x] Add release instructions file.

AnnaShaleva committed:
.github: upgrade actions/checkout and actions/setup-go GA version

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
.github: add release binary building job

A part of #78. Do not use the original Geth CI scripts since they are
more for debian packages. Deb pkg support may be added in future if
needed. Also, upload only Geth binary for now.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
.github: add Docker image build workflow

Close #103.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
Here are the notes from my side:
1. I tried to reuse the original https://github.com/bane-labs/go-ethereum/blob/bane-main/build/ci.go for Docker image build, but it complicates things related to Geth version a lot. This script is bound to `master` and `v1.` prefix of Geth version (but it's not a real problem). The problem is that this script can't properly get Git tag and branch name, so I wasn't able to manage this and thus I'm falling back to the GA job that we have in NeoGo.
2. We need to decide about versioning. I suggest to start our own versioning from some 0.1.0 because otherwise we may end up in a situation when original Geth release versions will mix with Bane versions. I also suggest to include the original Geth version to release notes of every Bane release.

AnnaShaleva commented on .github/workflows/build.yml:
```diff
@@ -12,34 +12,117 @@ on:
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to DockerHub
+        #if: ${{ github.event_name == 'release' || (github.event_name == 'workflow_dispatch' && github.event.inputs.push_image == 'true') }}
```
To be removed.

roman-khimov replied:
You want to push every build?

AnnaShaleva replied:
No, I'd like to push every release/manual build. This line should be uncommented.

roman-khimov replied:
OK, I thought "To be removed" means to remove the line.

AnnaShaleva commented:
So @roman-khimov, what do you think about versioning? Is it OK to start with 0.1.0?

roman-khimov commented:
I don't have any objections to that. We're certainly not 1.0 yet and we're not tied to geth releases at the same time.

AnnaShaleva committed:
params: begin v0.1.0 cycle

Ref.
https://github.com/ethereum/go-ethereum/commit/57d2b552c74dbd03b9909e6b8cd7b3de1f8b40e9.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
*: add release instructions and changelog

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
.github: enable macOS binaries builds

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
.github: set COMMIT Docker build arg for proper image labelling

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
Here's the test release example:  https://github.com/AnnaShaleva/go-ethereum/releases/tag/v0.1.0
The example of release job: https://github.com/AnnaShaleva/go-ethereum/actions/runs/8109348458
The example of version output got from released binary automatically attached to the GitHub release:
```
anna@kiwi:~/Downloads$ chmod +x geth-linux-amd64
anna@kiwi:~/Downloads$ ./geth-linux-amd64 --version
geth-linux-amd64 version 0.1.0-stable-9a72f319
anna@kiwi:~/Downloads$ ./geth-linux-amd64 version
Geth
Version: 0.1.0-stable
Git Commit: 9a72f3193c3463fe74183ec3b5f545e2e8ab738c
Architecture: amd64
Go Version: go1.21.7
Operating System: linux
GOPATH=
GOROOT=
```
The example of released Docker images automatically uploaded to the Dockerhub, two images for release 0.1.0: [stable](https://hub.docker.com/layers/annashaleva/geth/0.1.0/images/sha256-82a3ed2439e222f7eb7082e5dacbde28912bf060f4331a7c5decd4847f1780c7?context=repo) and [latest](https://hub.docker.com/layers/annashaleva/geth/latest/images/sha256-462eac2562409768531922386d306ed1a5a2c784bc7c68c0df2ce7c397dde350?context=repo).
Then version update is pushed to the example master and Build job is run manually: https://github.com/AnnaShaleva/go-ethereum/actions/runs/8109439783
The example of manually-triggered build of Docker image automatically uploaded to Dockerhub after release (single image without `latest` tag and marked as `unstable`): https://hub.docker.com/layers/annashaleva/geth/0.1.1-unstable-8930e10a-20240301/images/sha256-0d0ba9214e25ccbd52466444065e4a548245b88abe602d1a32ccfab8ec18d831?context=repo

AnnaShaleva committed:
.github: add Alltools Docker image build

Use the same rules as for Geth Docker image.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@roman-khimov, ready for review. Also, have minor updates that not yet tested (a couple of last commits), so it would be nice to wait until all jobs are finished.

AnnaShaleva committed:
.github: remove Windows builds support

Likely won't be added, but if needed, then may be requested in a
separate issue. Not supporting this for now.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
.github: add Alltools builder/uploader

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

roman-khimov commented on docs/release-instruction.md:
```diff
@@ -0,0 +1,87 @@
+
+Copy the GitHub release page link to:
+* Discord channel
+* Element channel
```
Not likely.

AnnaShaleva replied:
But why? We're going to open this repo someday, and it would be nice to share the release link at least in NeoGo Element channel (and Bane discord).

AnnaShaleva commented:
Tests are passed, let's merge and then I'll run action manually.
